accepted/2023/wasm-browser-threads.md

```diff
@@ -0,0 +1,682 @@
+- this thread needs to throw on any `.Wait` because of the problem **7**
+
+## HTTP and WS clients
+- are implemented in terms of `JSObject` and `Promise` proxies
+- they have thread affinity, see above
+    - typically to the `JSWebWorker` of the creator
+- but are consumed via their C# Streams from any thread.
+    - therefore need to solve the dispatch to correct thread.
+        - such dispatch will come with overhead
+        - especially when called with small buffer in tight loop
+    - or we could throw PNSE, but it may be difficult for user code to
+        - know what thread created the client
+        - have means how to dispatch the call there
+    - other unknowing users are `XmlUrlResolver`, `XmlDownloadManager`, `X509ResourceClient`, ...
+- because we could have blocking wait now, we could also implement synchronous APIs of HTTP/WS
+    - so that existing user code bases would just work without change
+    - this would also require separate thread, doing the async job
+    - we could use I/O thread for it
+
+## Performance
+As compared to ST build for dotnet wasm:
+- the dispatch between threads (caused by JS object thread affinity) will have negative performance impact on the JS interop
+- in case of HTTP/WS clients used via Streams, it could be surprizing
+- browser performance is lower when working with SharedArrayBuffer
+- Mono performance is lower because there are GC safe-points and locks in the VM code
+- startup is slower because creation of WebWorker instances is slow
+- VFS access is slow because it's dispatched to UI thread
+- console output is slow because it's POSIX stream is dispatched to UI thread, call per line
+
+# State 2023 September
+ - we already ship MT version of the runtime in the wasm-tools workload.
+ - It's enabled by `<WasmEnableThreads>true</WasmEnableThreads>` and it requires COOP HTTP headers.
+ - It will serve extra file `dotnet.native.worker.js`.
+ - This will also start in Blazor project, but UI rendering would not work.
+ - we have pre-allocated pool of browser Web Workers which are mapped to pthread dynamically.
+ - we can configure pthread to keep running after synchronous thread_main finished. That's necessary to run any async tasks involving JavaScript interop.
+ - legacy interop has problems with GC boundaries.
+ - JSImport & JSExport work
+ - There is private JSSynchronizationContext implementation which is too synchronous
+ - There is draft of public C# API for creating JSWebWorker with JS interop. It must be dedicated un-managed resource, because we could not cleanup JS state created by user code.
+ - There is MT version of HTTP & WS clients, which could be called from any thread but it's also too synchronous implementation.
+ - Many unit tests fail on MT https://github.com/dotnet/runtime/pull/91536
+ - there are MT C# ref assemblies, which don't throw PNSE for MT build of the runtime for blocking APIs.
+
+# Alternatives - as considered 2023 Sep
```

Comment:
I'm not sure if the alternative designs we considered last year are useful to stay in this document for posterity.
I'm happy to delete anything below.
We experimented with many of those options.

Reply:
Maybe move them to a public gist and link to it from here?